torch_xla/amp/autocast_mode.py

```diff
@@ -26,13 +27,15 @@ def __init__(self,
     self._xla_device = xm.xla_device_hw(device)
     if self._xla_device == 'GPU':
       backend = 'cuda'
+      self._cuda_bfloat16 = False
       if dtype is None:
         dtype = torch.float16
       elif dtype == torch.bfloat16:
         if xr.is_bf16_supported() and not torch.cuda.is_available():
           # XLA:GPU with bfloat16 should run on `xla` backend
           # unless torch.autocast is compiled with cuda.
           backend = 'xla'
+          self._cuda_bfloat16 = True
```

Comment:
Does PyTorch not support bfloat16 when running on xla backend? 
I'm still confused when dtype is set to bfloat16. 
When torch.cuda.is_available() returns False, torch_xla uses the XLA:GPU backend. However, in this patch it seems to fall back to use the GPU backend. Similarly, when torch.cuda.is_available() returns True, torch_xla uses the GPU backend, but the dtype is still forced to float16 instead of bfloat16.

Reply:
Hi @baoleai good catch -- I meant to remove the else restriction. We don't need to force torch.float16 on cuda backend.

Reply:
Ok, let me go over -- let me know if I am missing anything. 

1. >"When torch.cuda.is_available() returns False, torch_xla uses the XLA:GPU backend. However, in this patch it seems to fall back to use the GPU backend."

When cuda is not available (`False`), we go into this code path
```
if xr.is_bf16_supported() and not torch.cuda.is_available():
          # XLA:GPU with bfloat16 should run on `xla` backend
          # unless torch.autocast is compiled with cuda.
          backend = 'xla'
          self._cuda_bfloat16 = True
```
We uses `xla` backend for autocast.

2. >Similarly, when torch.cuda.is_available() returns True, torch_xla uses the GPU backend, but the dtype is still forced to float16 instead of bfloat16.

This is the part I need to address, instead of `else` we want `elif {cuda is not available}`, to implement the intention laid out in the err msg.

Reply:
So something like this,
```
if dtype is None:
        dtype = torch.float16
      elif dtype == torch.bfloat16 and not torch.cuda.is_available():
        if xr.is_bf16_supported():
          # XLA:GPU with bfloat16 should run on `xla` backend
          # unless torch.autocast is compiled with cuda.
          backend = 'xla'
          self._cuda_bfloat16 = True
        else:
          # This has been the default behavior for unsupported bfloat16 dtype
          dtype = torch.float16
          error_message = "In XLA:GPU autocast, but bfloat16 is not supported on this HW.\n"
          error_message += ("Using the default cuda autocast dtype float16.")
```